### PR TITLE
Add backup save/s3 delete backoff/retry and handle misconfigured restore

### PR DESCRIPTION
**Pull request by moelsayed**
- Add backoff for backup save
- Add retries for s3 delete
- Handle cases where restore is done with invalid credentials. 

https://github.com/rancher/rancher/issues/18484

### CODE REVIEW

File: pkg/controllers/management/clusterprovisioner/provisioner.go [hunk 1]
```diff
@@ -726,5 +726,21 @@ func (p *Provisioner) restoreClusterBackup(cluster *v3.Cluster, spec v3.ClusterS
...
 	}
-	return p.driverRestore(cluster, spec)
+
+	api, token, cert, err = p.driverRestore(cluster, spec)
+	if err != nil {
+		return "", "", "", err
+	}
+	// checking if we have s3 config and that it's not inconsistent. This happens
+	// when restore is performed with invalid credentials and then the cluster is updated to fix it.
+	if spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig.S3BackupConfig != nil {
```
Comment (alena1108) [inline pkg/controllers/management/clusterprovisioner/provisioner.go]: wouldn't error be not nil in this case? `when restore is performed with invalid credentials and then the cluster is updated to fix it.`

File: pkg/controllers/management/etcdbackup/etcdbackup.go [hunk 2]
```diff
@@ -112,7 +109,18 @@ func (c *Controller) Remove(b *v3.EtcdBackup) (runtime.Object, error) {
 	if b.Spec.BackupConfig.S3BackupConfig == nil {
 		return b, nil
 	}
-	return b, c.deleteS3Snapshot(b)
+	// try to remove from s3 for 3 times, then give up. if we don't we get stuck forever
+	var delErr error
+	for i := 0; i < 3; i++ {
+		if delErr := c.deleteS3Snapshot(b); delErr != nil {
```
Comment (kinarashah) [inline pkg/controllers/management/etcdbackup/etcdbackup.go]: I think we should omit if err is NotFound

File: pkg/controllers/management/etcdbackup/etcdbackup.go [hunk 3]
```diff
@@ -112,7 +109,18 @@ func (c *Controller) Remove(b *v3.EtcdBackup) (runtime.Object, error) {
...
+			logrus.Warnf("failed to delete backup from s3: %v", delErr)
+		}
+		time.Sleep(2 * time.Second)
```
Comment (kinarashah) [inline pkg/controllers/management/etcdbackup/etcdbackup.go]: need to break if delErr is nil

File: pkg/controllers/management/etcdbackup/etcdbackup.go [hunk 4]
```diff
@@ -112,7 +109,18 @@ func (c *Controller) Remove(b *v3.EtcdBackup) (runtime.Object, error) {
...
+	}
+	if delErr != nil {
+		logrus.Warnf("giving up on deleting backup [%s]: %v", b.Name, delErr)
+	}
+	return b, nil
```
Comment (kinarashah) [inline pkg/controllers/management/etcdbackup/etcdbackup.go]: just want to understand why are we returning nil here? we should return err here so controller retries

File: pkg/controllers/management/etcdbackup/etcdbackup.go [hunk 5]
```diff
@@ -192,6 +200,36 @@ func (c *Controller) createNewBackup(cluster *v3.Cluster) (*v3.EtcdBackup, error
...
+	}
+
+	return v3.BackupConditionCompleted.Do(b, func() (runtime.Object, error) {
+		cluster, err := c.clusterClient.Get(b.Spec.ClusterID, metav1.GetOptions{})
+		if err != nil {
+			return b, err
+		}
+		var inErr error
+		err = wait.ExponentialBackoff(backoff, func() (bool, error) {
+			if inErr = c.backupDriver.ETCDSave(context.Background(), cluster.Name, kontainerDriver, cluster.Spec, b.Name); err != inErr {
```
Comment (kinarashah) [inline pkg/controllers/management/etcdbackup/etcdbackup.go]: I want to understand why are we retrying the whole ETCDSave function? Are there some errors that this function can return for which we can just return?

File: pkg/controllers/management/clusterprovisioner/provisioner.go (hunk 1 above)
Comment (moelsayed) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: If the credentials are invalid, err will be nil. Once the cluster is updated with the correct credentials, the provisioner will kickin again, but we will still land in the restore case, not the update case (line 285). 

In this case the restore will complete successfully and error will be nil. But we still need to update the cluster to correct the credentials in the cluster store used by KE.

File: pkg/controllers/management/etcdbackup/etcdbackup.go (hunk 4 above)
Comment (moelsayed) [reply pkg/controllers/management/etcdbackup/etcdbackup.go]: If I return err the controller will be suck forever tryin to delete a bad backup. I retry to delete and when I fail I just give up.

File: pkg/controllers/management/etcdbackup/etcdbackup.go (hunk 2 above)
Comment (moelsayed) [reply pkg/controllers/management/etcdbackup/etcdbackup.go]: It's handled internally (or not) by the s3 library:
```
	// DeleteObject always responds with http '204' even for
	// objects which do not exist. So no need to handle them
	// specifically.
        return nil
```

File: pkg/controllers/management/etcdbackup/etcdbackup.go (hunk 4 above)
Comment (alena1108) [reply pkg/controllers/management/etcdbackup/etcdbackup.go]: This is fine for now. We follow the same pattern for the node provisioned by rancher. We retry removal on the remote provider several times, and then if it fails, we mark the node as removed anyway.

File: pkg/controllers/management/etcdbackup/etcdbackup.go (hunk 5 above)
Comment (kinarashah) [reply pkg/controllers/management/etcdbackup/etcdbackup.go]: Let's use context of the controller instead of background context

**Pull request merged**